d3a_interface/sim_results/kpi.py

```diff
@@ -145,11 +146,104 @@ def update_area_kpi(self, area_dict, core_stats):
         self._accumulate_energy_trace(core_stats)
 
 
+class SavingsKPI:
+    """Responsible for generating comparative savings from feed-in tariff vs D3A"""
+    def __init__(self):
+        self.producer_ess_set = set()  # keeps set of house's producing/ess devices
+        self.consumer_ess_set = set()  # keeps set of house's consuming/ess devices
+        self.fit_revenue = 0.  # revenue achieved by producing selling energy via FIT scheme
+        self.utility_bill = 0.  # cost of energy purchase from energy supplier
+        self.base_case_cost = 0.  # standard cost of a house participating in FIT scheme
+        self.d3a_cost = 0.  # standard cost of a house participating in D3A
+        self.saving_absolute = 0.  # savings achieved by a house via participating in D3A
+        self.saving_percentage = 0.  # savings in percentage wrt FIT via participating in D3A
+
+    def calculate_savings_kpi(self, area_dict: dict, core_stats: dict, gf_alp: float):
+        """Calculates the referenced saving from feed-in tariff based participation vs D3A
+        Args:
+            area_dict: contain nested area info
+            core_stats: contain area"s raw/key statistics
+            gf_alp: grid_fee_along_the_path - cumulative grid fee from root to target area
+
+        """
+        self.populate_consumer_producer_sets(area_dict)
+
+        # fir_excl_gf_alp: feed-in tariff rate excluding grid fee along path
+        fir_excl_gf_alp = self.get_feed_in_tariff_rate_excluding_path_grid_fees(
+            core_stats.get(area_dict["uuid"], {}), gf_alp)
+        # mmr_incl_gf_alp: market maker rate include grid fee along path
+        mmr_incl_gf_alp = self.get_market_maker_rate_including_path_grid_fees(
+            core_stats.get(area_dict["uuid"], {}), gf_alp)
+        for trade in core_stats.get(area_dict["uuid"], {}).get("trades", []):
+            if trade["seller_origin_id"] in self.producer_ess_set:
+                self.fit_revenue += fir_excl_gf_alp * trade["energy"]
+                self.d3a_cost -= trade["price"]
+            if trade["buyer_origin_id"] in self.consumer_ess_set and \
+                    trade["seller_origin_id"] not in self.producer_ess_set:
+                self.utility_bill += mmr_incl_gf_alp * trade["energy"]
+                self.d3a_cost += trade["price"]
+        self.base_case_cost = self.utility_bill - self.fit_revenue
+        self.saving_absolute = self.base_case_cost - self.d3a_cost
+        self.saving_percentage = ((self.saving_absolute / self.base_case_cost) * 100
+                                  if self.base_case_cost else 0.)
+
+    def populate_consumer_producer_sets(self, area_dict: dict):
+        """
+        Responsible for devices' classification
```

Comment:
Docstring should be written in imperative and should be ending with a dot.
```suggestion
        Populate sets of device classes.
```